### PR TITLE
Use GitHub Releases for official builds 

### PR DESCRIPTION
**Pull request by Thog**
This get ride of AppVeyor for official builds once and for all.

As people have been using AppVeyor for custom build and as number might clash with official releases, the minor version of the project was incremented.

AppVeyor will still be used for PRs.

Also to avoid more confusion between AppVeyor builds and official one, AppVeyor builds are now postfixed with "-prXXX".

### CODE REVIEW

**Comment by jduncanator** [comment]
What is the maintenance plan for AppVeyor moving forward? Seems like this will require twice the work to maintain if we ever add new build requirements or features, as the changes will need to be done against both GitHub Actions and AppVeyor.

Is GitHub Actions not capable of building the PRs also? What was the reasoning behind moving away from AppVeyor? It seems, at least from what I've seen and heard, that GitHub Actions is a less capable and featured solution, what are some of the merits of moving to GitHub Actions?

**Comment by Thog** [comment]
@jduncanator 

GitHub Actions is already building PRs, the only missing feature is an artifacts storage. (the feature is here but limited to 250MB per repository)

The major reason behind moving away from AppVeyor for master building is to have proper releases, not build artifacts. AppVeyor isn't designed to be a long term storage for **releases** as every builds expires after 6 months. Add to this that AppVeyor sometime silently generating wrong artifacts and that runner takes quite some time to start (already had to wait 15 min for a build to start) and you have my reasons.


PS: If you want to unify the workflow, we could probably use MSBuild task as you said in private or we can also do docker images like RPCS3. We could also consider moving to Azure Pipeline.

File: Ryujinx.Actions/ArtifactInformation.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
+﻿using System.IO;
+using System.Security.Cryptography;
+
+namespace Ryujinx.Actions
+{
+    public class ArtifactInformation
+    {
+        public OperatingSystem Os  { get; private set; }
```
Comment (AcK77) [inline Ryujinx.Actions/ArtifactInformation.cs]: Don't forget to align names if you want to align `{}`

File: Ryujinx.Actions/ArtifactInformation.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        public BuildType BuildType { get; private set; }
+        public string Url          { get; set; }
+        public string FileHash     { get; private set; }
+        public string FileName     { get; private set; }
+
+        private static OperatingSystem GetOperatingSystemFromNamingPart(string part)
+        {
+            switch (part)
+            {
+                case "win":
```
Comment (AcK77) [inline Ryujinx.Actions/ArtifactInformation.cs]: Maybe you can put `return` on the same line and align them?

File: Ryujinx.Actions/ArtifactInformation.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                default:
+                    return OperatingSystem.Unknown;
+            }
+        }
+
+        private static Architecture GetArchitectureFromNamingPart(string part)
+        {
+            switch (part)
+            {
+                case "x64":
```
Comment (AcK77) [inline Ryujinx.Actions/ArtifactInformation.cs]: Maybe you can put `return` on the same line and align them?

File: Ryujinx.Actions/ArtifactInformation.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                case "aarch64":
+                    return Architecture.AArch64;
+                default:
+                    return Architecture.Unknown;
+            }
+        }
+
+        public static ArtifactInformation FromFileName(string fileName)
+        {
+            string fileNameWithoutExtension = Path.GetFileNameWithoutExtension(fileName);
```
Comment (AcK77) [inline Ryujinx.Actions/ArtifactInformation.cs]: Sometime you align assignment, sometime you don't align name, and sometime you align everything.
You should make a choice :P

File: Ryujinx.Actions/ArtifactInformation.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            artifactInformation.BuildType = BuildType.Release;
+
+            bool isProfile = fileNameParts.Length == 4;
+
+            if (isProfile)
+            {
+                artifactInformation.BuildType = BuildType.ProfileRelease;
+            }
+
+            string targetConfiguration        = isProfile ? fileNameParts[3] : fileNameParts[2];
```
Comment (AcK77) [inline Ryujinx.Actions/ArtifactInformation.cs]: Align names if you align assigns

File: Ryujinx.Actions/Program.cs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            client.Credentials = new Credentials(releaseToken);
+
+            return client;
+        }
+
+        private static JsonSerializerOptions CreateJsonOptions()
+        {
+            JsonSerializerOptions serializeOptions = new JsonSerializerOptions
+            {
+                IgnoreNullValues = true,
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]: You align assignments above in `assetUpload` so you should align them here too

File: Ryujinx.Actions/Program.cs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                await UploadAsset(client, release, fileStream, artifactInformation.FileName, "application/zip");
+            }
+        }
+
+        private static async Task PublishRelease(string artifactDirectory, VersionCore releaseVersion, string releaseRepository, string releaseToken)
+        {
+            if (releaseVersion == null)
+            {
+                Core.SetFailed("Invalid release_version!");
+                return;
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]: Add a new empty line before a `return` (same for all conditions above)

File: Ryujinx.Actions/Program.cs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            }
+
+            if (releaseRepository == null)
+            {
+                Core.SetFailed("Invalid release_repository!");
+                return;
+            }
+
+            string commitSha = Environment.GetEnvironmentVariable("GITHUB_SHA");
+
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]: Extra empty line maybe?

File: Ryujinx.Actions/Program.cs [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            if (commitSha == null)
+            {
+                Core.SetFailed("GITHUB_SHA environment variable not found!");
+                return;
+            }
+
+            string[] releaseRepositoryParts = releaseRepository.Split("/");
+
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]: Extra empty line maybe?

File: Ryujinx.Actions/Program.cs [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            {
+                ArtifactInformation artifactInformation = ArtifactInformation.FromFile(file);
+
+                releaseInformation.Artifacts.Add(artifactInformation);
+            }
+
+            GitHubClient client = CreateGitHubClientInstance(releaseToken);
+
+            // Create a new release
+            NewRelease newRelease = new NewRelease($"build-{commitSha}");
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]:
```csharp
NewRelease newRelease = new NewRelease($"build-{commitSha}")
{
     Name = blabla,
     Body = blabla,
     etc...
};
```

File: Ryujinx.Actions/Program.cs [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+            newRelease.Name       = releaseVersion.ToString();
+            newRelease.Body       = $"Triggered by {GITHUB_COMMIT_BASE_URL}{commitSha}.";
+            newRelease.Draft      = true;
+            newRelease.Prerelease = false;
+
+            Release release             = await client.Repository.Release.Create(releaseRepositoryOwner, releaseRepositoryName, newRelease);
```
Comment (AcK77) [inline Ryujinx.Actions/Program.cs]: Align names

File: Ryujinx.Actions/ReleaseInformation.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+using System.Text;
+
+namespace Ryujinx.Actions
+{
+    public class ReleaseInformation
+    {
+        public VersionCore Version { get; set; }
+        public List<ArtifactInformation> Artifacts { get; set; }
+
+        private ReleaseInformation()
```
Comment (AcK77) [inline Ryujinx.Actions/ReleaseInformation.cs]:
```suggestion
        private ReleaseInformation() { }
```

**Comment by jduncanator** [comment]
@Thog 

That's valid, however AppVeyor can deploy artifacts to GitHub Releases as a post-build process, so if you wanted to have a more solid release process, that would have been a solution also. 

I wasn't aware that we were having any issues with AppVeyor builds generating incorrect artifacts, that sounds more like a misconfiguration than an issue with AppVeyor to me.

I'm open to moving build and CI providers, I just want to make sure we're doing it for the right reasons, we understand what those reasons and issues are, and we make an informed decision on what the best platform is going forward. To me, personally, GitHub Actions is not ready for prime time yet, and is missing too many features and has too many limitations to be a primary choice for CI, but without understanding what our requirements are, we can't really make a decision on the best solution.

For now, this looks fine, however I don't really like the idea of splitting PRs and releases across two different build platforms. I also don't think the Ryujinx.Actions project belongs in the main Ryujinx repository. I think it would be best to create another repository and store that code there, it doesn't really have anything to do with Ryujinx, and just adds clutter to the already cluttered repo.

**Comment by Thealexbarney** [comment]
Using GitHub releases for storing every build of master seems like it would get cluttered rather quickly. In general I'd assume that releases were manually selected instead of done for every commit to master.

What do you mean by "incorrect artifacts", @Thog? Ideally you'd be able to run the build script locally while only using CI-Provider-specific stuff for things like posting artifacts or specifying the build configurations that should be built.

**Comment by Margen67** [comment]
> 
> 
> Using GitHub releases for storing every build of master seems like it would get cluttered rather quickly. In general I'd assume that releases were manually selected instead of done for every commit to master.
> 

rpcs3 works around this by having a separate repository for releases: https://github.com/RPCS3/rpcs3-binaries-win
#
You can link to the latest artifacts like AppVeyor lets you do by making a link to the latest release: https://docs.github.com/en/github/administering-a-repository/linking-to-releases

**Comment by Thog** [comment]
Closing for now, I plan to redo it later.

As @Margen67 said, storing GitHub Releases for every commit is entirely fine and is already being used by RPCS3 so I will keep most of the current design but add into consideration multiple channel releases.

**Pull request closed without merging**